### PR TITLE
Build add chassis form.

### LINKED ISSUE #817: Write http method for adding chassis
No websocket method currently. LP bug: https://bugs.launchpad.net/maas/+bug/1863906

### PR DESCRIPTION
**Pull request by Caleb-Ellis**
## Done
- Added machine addChassis http method and saga/actions/reducers. LP bug about websocket implementation here: https://bugs.launchpad.net/maas/+bug/1863906
- Built AddChassisForm component.
- Refactored the AddMachineForm to use some common hooks and components.
- Note that chassis types are currently hardcoded, as in the legacy version. Relevant LP bug here: https://bugs.launchpad.net/maas/+bug/1863905

## QA
- Click the Add hardware dropdown and select Chassis.
- Check that you can change power type and the power parameters fields change.
- Check form validation works as expected.
- Enter a hostname/address that your MAAS is aware of (in my case I used karura's only KVM's virsh address) and submit.
- Check that a spinner shows and then you are redirected to the machine list with a confirmation notification.
- Enter incorrect data and submit and check that you get an error notification above the form.
- Check that you can still add machines in the Add machine form

## Fixes
Fixes canonical-web-and-design/MAAS-squad#1730
Fixes #817 

### Default
0 0 0 0_8400_MAAS_r_machines_chassis_add (1)

### Form validation
0 0 0 0_8400_MAAS_r_machines_chassis_add (2)

### Server error
0 0 0 0_8400_MAAS_r_machines_chassis_add (3)

### Confirmation notification
0 0 0 0_8400_MAAS_r_machines_chassis_add (5)

### CODE REVIEW

**Comment by huwshimi** [comment]
This has a conflict now (probably with my last branch sorry).

**Comment by squidsoup** [comment]
There seems to be some weird styling on "Power Control"

image

File: ui/src/app/base/hooks.js [hunk 1]
```diff
@@ -214,3 +215,53 @@ export const useMachineActions = (systemId, actions, noneMessage, onClick) => {
...
+      powerTypes.reduce((parameters, powerType) => {
+        powerType.fields.forEach(field => {
+          if (!(field.name in parameters)) {
+            parameters[field.name] = field.default;
+          }
+        });
+        return parameters;
+      }, {}),
+    [powerTypes]
+  );
```
Comment (squidsoup) [inline ui/src/app/base/hooks.js]: Nice

File: ui/src/app/machines/views/AddChassis/chassisPowerTypes.js [hunk 2]
```diff
@@ -0,0 +1,247 @@
+/**
+ * Hard coded chassis types. This is because there is currently no method in
+ * MAAS to get a full list of supported chassis. Remove this file when this bug
+ * has been fixed: https://bugs.launchpad.net/maas/+bug/1863905
```
Comment (squidsoup) [inline ui/src/app/machines/views/AddChassis/chassisPowerTypes.js]: thanks for filing the bug

**Comment by squidsoup** [comment]
@Caleb-Ellis apologies, some conflicts to fix again.

**Comment by anthonydillon** [comment]
@sparkiegeek do you know if there is a way we can test this for real? Feels risky to build UI without a reall way of testing it?

**Comment by Caleb-Ellis** [comment]
> There seems to be some weird styling on "Power Control"
> 
> image

That's odd, I'm not getting the same styling issue on firefox or chrome. Seems to only happen with the select. @anthonydillon are you able to take a look and see if you see anything weird?

> @sparkiegeek do you know if there is a way we can test this for real? Feels risky to build UI without a reall way of testing it?

With this PR it's possible to get a 200 from the backend if we make it happy with the right form values, but as for actually "adding a chassis"... we should probably test that for real.

**Comment by Caleb-Ellis** [comment]
Going to merge this now. Any styling fixes can be added to #794 or #795 because I'll be touching the same files.

**Pull request merged**